docs/src/main/asciidoc/security-openid-connect-multitenancy-concept.adoc

```diff
@@ -0,0 +1,317 @@
+  // In real implementations, only 1 is used.
+
+  @Inject
+  protected lateinit var accessToken: JsonWebToken
+
+  @Inject
+  @IdToken
+  protected lateinit var idToken: JsonWebToken
+
+  @Inject
+  protected lateinit var callerInfo: TokenBasedCallerInfo
+}
+----
+
+A custom wrapper for the injected Access Token is often preferrable to using the JsonWebToken directly, since the JsonWebToken API retrieves all claims by key, similar to a Map. In enterprise environments, it is frequently desirable to reduce the potential for spelling errors or other mistakes when accessing the JsonWebToken claims - hence injecting your own `@RequestScoped` wrapper around the Access Token will frequently yield stabler and simpler results.
+
+The `TokenBasedCallerInfo` below shows the structure for a custom wrapper around the injected Access Token:
+
+[source,kotlin]
+----
+@RequestScoped
+@UseOpenMembers
+class TokenBasedCallerInfo  {
+
+    @Inject
+    protected lateinit var accessToken: JsonWebToken
+
+    @PostConstruct
+    fun onConstruction() {
+
+        if (log.isDebugEnabled) {
+            log.debug("Got AccessToken: $accessToken")
+        }
+    }
+
+    // Usability-enhancing getters
+    // Expose values parsed from the injected JsonWebToken in a user-friendly manner
+
+    val firstName: String by lazy { accessToken.getClaim("given_name") }
+    val lastName: String by lazy { accessToken.getClaim("family_name") }
+    val email: String by lazy { accessToken.getClaim("email") }
+    val roles: SortedSet<String> by lazy { rolesFromRealmAccess(accessToken) }
+    val userId: Long
+        get() {
+            // Find the userId within the accessToken
+        }
+
+    // ... more getters, as required ...
+
+    companion object {
+
+        @JvmStatic
+        fun rolesFromRealmAccess(accessToken: JsonWebToken): SortedSet<String> {
+            // Parse the Access Token to find given roles
+        }
+
+        // ... more methods to parse the Access Token for custom purposes ...
+
+        @JvmStatic
+        private val log = LoggerFactory.getLogger(TokenBasedCallerInfo::class.java)
+    }
+}
+----
+
+This implementation concludes step `2` in Figure 2 above. We have implemented a wrapper around the AccessToken available from Quarkus' security framework, and injected user information (i.e. JsonWebToken claims) into the multi-tenant application.
+
+=== 3. Validate user authorization to retrieve or create data.
+
+Multi-tenant applications of some complexity tend to require both annotated and programmatic authorization for users calling resource methods. Since we now have user information injected (assuming we chose to use the `TokenBasedCallerInfo` request scoped bean defined above), we can use its properties in any `AbstractCalendarResource` subclass - and any other class in which we inject the `TokenBasedCallerInfo`. The common/shared information about the service's resource path and used roles is extracted to a shared singleton holding constants:
```

Comment:
For the concept doc like this one, I'd not base it on the custom token info example, it complicates things IMHO